c10/core/SingletonSymNodeImpl.h

```diff
@@ -81,42 +108,72 @@ class C10_API SingletonSymNodeImpl : public SymNodeImpl {
     return SymNode(c10::make_intrusive<ConstantSymNodeImpl<bool>>(ret));
   }
 
-  // It would be cool to have the ability to arbitrarily constrain the range of
-  // values as we do for unbacked symints. For now a useful default
-  // range seems to be [2, int64_t::max()] (1) since sizes are non-negative, and
-  // (2) we need to get past 0/1 specialization checks.
   c10::SymNode ge(const c10::SymNode& other) override {
-    if (auto mb_si = other->singleton_int()) {
-      return SymNode(
-          c10::make_intrusive<ConstantSymNodeImpl<bool>>(val_ == *mb_si));
+    auto mb_si = other->singleton_int();
+    if ((mb_si.has_value() && val_ == *mb_si) ||
+        (other->constant_int() && *other->constant_int() <= 2)) {
+      return SymNode(c10::make_intrusive<ConstantSymNodeImpl<bool>>(true));
     }
-    c10::optional<int64_t> c = other->constant_int();
-    TORCH_CHECK(c.has_value());
-    return SymNode(c10::make_intrusive<ConstantSymNodeImpl<bool>>(*c <= 2));
+    TORCH_CHECK(
+        false,
+        "singleton ge: '",
+        this->str(),
+        " >= ",
+        (mb_si.has_value() ? *mb_si : *other->constant_int()),
+        "' is indeterminate");
   }
 
   c10::SymNode gt(const c10::SymNode& other) override {
-    if (auto mb_si = other->singleton_int()) {
+    auto mb_si = other->singleton_int();
+    if (mb_si.has_value() && val_ == *mb_si) {
       return SymNode(c10::make_intrusive<ConstantSymNodeImpl<bool>>(false));
+    } else if (other->constant_int() && *other->constant_int() < 2) {
+      return SymNode(c10::make_intrusive<ConstantSymNodeImpl<bool>>(true));
     }
-    c10::optional<int64_t> c = other->constant_int();
-    TORCH_CHECK(c.has_value());
-    return SymNode(c10::make_intrusive<ConstantSymNodeImpl<bool>>(*c < 2));
+    TORCH_CHECK(
+        false,
+        "singleton gt: '",
+        this->str(),
+        " > ",
+        (mb_si.has_value() ? *mb_si : *other->constant_int()),
+        "' is indeterminate");
   }
 
   c10::SymNode lt(const c10::SymNode& other) override {
-    return SymNode(c10::make_intrusive<ConstantSymNodeImpl<bool>>(false));
+    auto mb_si = other->singleton_int();
+    if (mb_si.has_value() && val_ == *mb_si) {
+      return SymNode(c10::make_intrusive<ConstantSymNodeImpl<bool>>(false));
+    } else if (other->constant_int() && *other->constant_int() < 2) {
+      return SymNode(c10::make_intrusive<ConstantSymNodeImpl<bool>>(false));
+    }
+    TORCH_CHECK(
+        false,
+        "singleton lt: '",
+        this->str(),
+        " < ",
+        (mb_si.has_value() ? *mb_si : *other->constant_int()),
+        "' is indeterminate");
   }
 
   c10::SymNode le(const c10::SymNode& other) override {
-    if (auto mb_si = other->singleton_int()) {
-      return SymNode(
-          c10::make_intrusive<ConstantSymNodeImpl<bool>>(val_ == *mb_si));
+    auto mb_si = other->singleton_int();
```

Comment:
You should seriously consider just defining these in terms of their negations, reduce boilerplate

Reply:
This seems tricky to do because there are a lot of exceptions to the rule. 

| Operation | `< 2`  | `2` | `> 2 but < int64_t::max` | `int64_t::max` |
|-----------|-------|-----|--------------------------|----------------|
| `ge`      | `true`| `true` | Error (Indeterminate) | Error (Indeterminate) |
| `le`      | `false`| Error (Indeterminate) | Error (Indeterminate) | `true` |

Reply:
With the semantics as-is there are too many edge cases for this to actually be very simplifying:

| Operation | `< 2`  | `2` | `> 2 but < int64_t::max` | `int64_t::max` |
|-----------|-------|-----|--------------------------|----------------|
| `ge`      | `true`| `true` | Indeterminate | Indeterminate | 
| `le`      | `false`| Indeterminate | Indeterminate | `true` |

However, it might actually be reasonable just to make `j <= 2` be `false` and `j > 2` be `true`.

Reply:
Ignore my previous message, it is indeed possible to simplifying as-is.